Factor out certificate renewal logic out of hosted service

dv00d00 opened the pull request:
This is a quite big PR.

Main goals:

- extract core domain logic to allow future unit tests
- allow a user to control the key size and thus generated certificate size via a new property on `LetsEncryptOptions`

while I was moving things around, the reduced code size in the `LetsEncryptRenewalService` allowed to spot a potential bug. I've noticed _timer field in `LetsEncryptRenewalService` was not always accessed without null guards and have fixed those as well

I've tried to maintain binary compatibility, so this should not be a breaking change

dv00d00 commented:
Spotted an issue with `ICertificateRenewalLifecycleHook.OnRenewalSucceededAsync` was not called, fixed it as well.

I believe PR is ready for review now. Not sure but it might be worth splitting it into several pieces.

ffMathy commented on src/FluffySpoon.AspNet.LetsEncrypt/CertificateRenewal.cs:
```diff
@@ -0,0 +1,221 @@
+			_options = options;
+		}
+
+        public async Task<(X509Certificate2,CertificateRenewalStatus)> RenewCertificateIfNeeded(X509Certificate2 current)
```
Instead of a Tuple here, can we use a struct or class? It's always more readable in my opinion.

ffMathy replied:
@dv00d00 did you see this?

dv00d00 replied:
I've replaced it with `CertificateRenewalResult` class

ffMathy replied:
Ah didn't catch that - thanks.

ffMathy commented:
Can you elaborate on why extracting it out in a separate class helps with testing?

dv00d00 commented:
Original LetsEncryptRenewalService was doing too much: (hosted service, timers, persistence, lifetime hooks, acme challenges, global mutable state)

Extraction of the core domain as in this pr separates acme challenges and persistence logic from the overloaded service. I was able to spot several issues just by reading the reduced amount of code ( _timer was accessed without null guards, certificate renewed lifetime hook was not invoked)

I've tried to make new service slightly more functional (e.g. do not mutate global state, just return the result of a computation). Reduced scope means reduced mock surface = easier setups. Potential next steps would be to abstract Certes via some sort of interfaces and we will have full control over the business logic just by mocking two dependencies (acme and persistence).

I understand my reasoning might be subjective, but elimination of a god class and reduction of complexity of a unit is surely beneficial in a long run

ffMathy commented:
Alright, perfect - that makes sense.

I just don't necessarily like the `ICertificateRenewal` name. Can we figure out another name for it?

dv00d00 commented:
> Alright, perfect - that makes sense.
> 
> I just don't necessarily like the `ICertificateRenewal` name. Can we figure out another name for it?

Naming is hard, I will think about a better name, but please feel free to suggest anything more appropriate.

dv00d00 commented:
I've renamed CertificateRenewal to LetsEncryptClient and moved it with its friends to the Logic namespace. `Logic` sounds wanky but it is the best name I have had come up with, suggestions are welcome.

ffMathy commented:
I like the `LetsEncryptClient` name - it feels correct at first glance.

However, the client also borrows the responsibility of persistence - seems a bit weird for a "client" to do so.

What if we moved everything LetsEncrypt related into the client, and then everything else back into the renewal service?

dv00d00 commented:
I've tried to do this today. This sort of refactoring produced ripples across core business logic, I would not dare to attempt such refactoring before writing a suite of unit tests first.

Factors:

- inherently OO design of certes forces propagation of state, making it difficult to decouple from perf/rate-limiting concerns of certificate acquisition
- middleware logic is temporally coupled to the certificate acquisition. If we won't store challenges before requesting http validation the validation process will fail
- the refactoring in this pr is pretty simple because it just moves a chunk of methods from one class to another, without modifying logic at all

Suggestion: 
- Rename LetsEncryptClient to LetsEncryptFacade or something similar
- Cover happy path and expected failures with unit tests
- Attempt major refactoring of decoupling certes out of persistence

dv00d00 commented:
demo pr: https://github.com/dv00d00/FluffySpoon.AspNet.LetsEncrypt/pull/1

ffMathy commented:
Very cool work @dv00d00 and thank you so much.

I'll merge. Would you be interested in looking at it after? 👍

ffMathy commented:
But first, you still haven't replied to the review comment.

dv00d00 commented:
I am planning to provide a few more pull requests (some sort of tests first, deeper refactoring after)
Due to the review, I've replaced touple with a named class

ffMathy commented:
Perfect!
